src/helpers/buffers/send.rs

```diff
@@ -1,61 +1,314 @@
+use crate::helpers::buffers::fsv::FixedSizeByteVec;
 use crate::helpers::fabric::{ChannelId, MessageEnvelope};
+use crate::protocol::RecordId;
 use std::collections::hash_map::Entry;
 use std::collections::HashMap;
-use std::mem;
+use std::ops::Range;
+
+/// Use the buffer that allocates 8 bytes per element. It could probably go down to 4 if the
+/// only thing IPA sends is a single field value. To support arbitrarily sized values, it needs
+/// to be at least 16 bytes to be able to store a fat pointer in it.
+type ByteBuf = FixedSizeByteVec<8>;
 
 /// Buffer that keeps messages that must be sent to other helpers
 #[derive(Debug)]
-pub(in crate::helpers) struct SendBuffer {
-    max_capacity: usize,
-    inner: HashMap<ChannelId, Vec<MessageEnvelope>>,
+#[allow(clippy::module_name_repetitions)]
+pub struct SendBuffer {
+    items_in_batch: usize,
+    batch_count: usize,
+    inner: HashMap<ChannelId, ByteBuf>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum PushError {
+    #[error("Record {record_id:?} is out of accepted range {accepted_range:?}")]
+    OutOfRange {
+        channel_id: ChannelId,
+        record_id: RecordId,
+        accepted_range: Range<RecordId>,
+    },
+    #[error(
+        "Message with the same record id {record_id:?} has been already sent to {channel_id:?}"
+    )]
+    Duplicate {
+        channel_id: ChannelId,
+        record_id: RecordId,
+        previous_value: Box<[u8]>,
+    },
+}
+
+/// Send buffer configuration is defined over two parameters. `items_in_batch` indicates how many
+/// elements a single request to send data to network layer can hold. The size of the batch in
+/// bytes is defined as `items_in_batch` * `ByteBuf::ELEMENT_SIZE_BYTES`.
+///
+/// `batch_count` defines the overall capacity of the send buffer as `items_in_batch` * `batch_count`.
+/// Setting it to `2` will double the capacity of the send buffer as at most two batches can be kept
+/// in memory, `4` quadruples the capacity etc.
+#[derive(Debug, Copy, Clone)]
+pub struct Config {
+    pub items_in_batch: u32,
+    pub batch_count: u32,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            items_in_batch: 1,
+            batch_count: 1,
+        }
+    }
 }
 
 impl SendBuffer {
-    pub fn new(max_channel_capacity: u32) -> Self {
+    pub fn new(config: Config) -> Self {
         Self {
-            max_capacity: max_channel_capacity as usize,
+            items_in_batch: config.items_in_batch as usize,
+            batch_count: config.batch_count as usize,
             inner: HashMap::default(),
         }
     }
 
+    /// TODO: change the output to Vec<u8> - we no longer need a wrapper. The raw byte vector
+    /// will be communicated down to the network layer.
+    #[allow(clippy::needless_pass_by_value)] // will be fixed when tiny/smallvec is used
     pub fn push(
         &mut self,
         channel_id: ChannelId,
```

Comment:
I think that this should take `Borrow<ChannelId>`, because you don't need to create a clone unless you are adding a new item to the map.